code/obj/machinery/plantpot.dm

```diff
@@ -140,44 +140,42 @@
 
 		frequency.post_signal(src, signal)
 
-	on_reagent_change()
-		src.update_water()
-		if (!current) // Make sure we don't have a plant in the tray first
-			var/current_total_volume = (src.reagents ? src.reagents.total_volume : 0)
-			var/current_water_level = (src.reagents ? src.reagents.get_reagent_amount("water") : 0)
-
+	proc/update_water_level() //checks reagent contents of the pot, then returns the cuurent water level
+		var/current_total_volume = (src.reagents ? src.reagents.total_volume : 0)
+		var/current_water_level = (src.reagents ? src.reagents.get_reagent_amount("water") : 0)
+		switch(current_water_level)
+			if(0) current_water_level = 1
+			if(1 to 40) current_water_level = 2
+			if(41 to 100) current_water_level = 3
+			if(101 to 200) current_water_level = 4
+			if(201 to INFINITY) current_water_level = 5
+		if(current_water_level != src.water_level)
+			src.water_level = current_water_level
+		if(!current) // Make sure we don't have a plant in the tray first (layering stuffs)
 			switch(current_total_volume)
-				if (0) current_total_volume = 1
-				if (1 to 40) current_total_volume = 2
-				if (41 to 100) current_total_volume = 3
-				if (101 to 200) current_total_volume = 4
-				if (201 to INFINITY) current_total_volume = 5
-
-			if (current_total_volume != src.total_volume)
+				if(0) current_total_volume = 1
+				if(1 to 40) current_total_volume = 2
+				if(41 to 100) current_total_volume = 3
+				if(101 to 200) current_total_volume = 4
+				if(201 to INFINITY) current_total_volume = 5
+			if(current_total_volume != src.total_volume)
 				src.total_volume = current_total_volume
-				do_update_icon = 1
 
-			switch(current_water_level)
-				if (0) current_water_level = 1
-				if (1 to 40) current_water_level = 2
-				if (41 to 100) current_water_level = 3
-				if (101 to 200) current_water_level = 4
-				if (201 to INFINITY) current_water_level = 5
+		src.update_water_icon()
```

Comment:
You'll want a check either before or in this proc. You don't want to create images and update overlays unnecessarily, but only when the overlay states have changed. Right now, if there are reagents in the pot while a plant is growing, `on_reagent_change()` will be called every `process()`, which will then proc `update_water_level()` and `update_water_icon()`. Similar to how there's a `do_update_icon`, you should probably have a `do_water_icon`.

Reply:
`reagents.remove_any_except()` > `reagents.remove_reagent()` > `reagents.reagents_changed()` > `my_atom.on_reagent_change()`